tests/validator/validator.cpp

```diff
@@ -2395,6 +2357,333 @@ TEST(Validator, unfoundUnitsInEncapsulatedComponents)
+{
+    const std::vector<std::string> expectedErrors = {
+        "Variable 'v1' has an interface type set to 'public' which is not the correct interface type for this variable. The interface type required is 'private'.",
+        "The equivalence between 'v2' in component 'c2'  and 'v4' in component 'c4' is invalid. Component 'c2' and 'c4' are neither siblings nor in a parent/child relationship.",
+        "The equivalence between 'v4' in component 'c4'  and 'v2' in component 'c2' is invalid. Component 'c4' and 'c2' are neither siblings nor in a parent/child relationship.",
+    };
+
+    libcellml::ModelPtr model = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+    libcellml::ComponentPtr c2 = libcellml::Component::create();
+    libcellml::ComponentPtr c3 = libcellml::Component::create();
+    libcellml::ComponentPtr c4 = libcellml::Component::create();
+    libcellml::ComponentPtr c5 = libcellml::Component::create();
+
+    libcellml::ValidatorPtr validator = libcellml::Validator::create();
+
+    model->setName("model");
+    c1->setName("c1");
+    c2->setName("c2");
+    c3->setName("c3");
+    c4->setName("c4");
+    c5->setName("c5");
+
+    model->addComponent(c1);
+    c1->addComponent(c2);
+    c1->addComponent(c3);
+    c2->addComponent(c5);
+    c3->addComponent(c4);
+
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+    v1->setName("v1");
+    v1->setUnits("dimensionless");
+    v1->setInterfaceType("public");
+
+    libcellml::VariablePtr v2 = libcellml::Variable::create();
+    v2->setName("v2");
+    v2->setUnits("dimensionless");
+    v2->setInterfaceType("public_and_private");
+
+    libcellml::VariablePtr v3 = libcellml::Variable::create();
+    v3->setName("v3");
+    v3->setUnits("dimensionless");
+    v3->setInterfaceType("public");
+
+    libcellml::VariablePtr v4 = libcellml::Variable::create();
+    v4->setName("v4");
+    v4->setUnits("dimensionless");
+
+    libcellml::VariablePtr v5 = libcellml::Variable::create();
+    v5->setName("v5");
+    v5->setUnits("dimensionless");
+    v5->setInterfaceType("public");
+
+    c1->addVariable(v1);
+    c2->addVariable(v2);
+    c3->addVariable(v3);
+    c4->addVariable(v4);
+    c5->addVariable(v5);
+
+    libcellml::Variable::addEquivalence(v2, v4); // invalid equivalence, too far away
+    libcellml::Variable::addEquivalence(v1, v2);
+    libcellml::Variable::addEquivalence(v2, v3);
+    libcellml::Variable::addEquivalence(v2, v5);
+
+    validator->validateModel(model);
+
+    EXPECT_EQ_ERRORS(expectedErrors, validator);
+}
+
+TEST(Validator, variableInterfaceShouldBePublic)
+{
+    const std::vector<std::string> expectedErrors = {
+        "Variable 'v1' has no interface type set. The interface type required is 'public'.",
+        "Variable 'v2' has an interface type set to 'private' which is not the correct interface type for this variable. The interface type required is 'public'.",
```

Comment:
Other cases of where I wish we would also be told about the problem, not only its solution.

Reply:
We can add more information to the informative spec around the causes and reasons - it will save having enormously long error messages returned from the API.